### PR TITLE
Refactoring for e2e tests

### PR DESCRIPTION
**Pull request by sbryzak**
This PR:

1. Introduces a new end to end deactivation lifecycle test
2. Refactors all tests to use registration service to create new UserSignup instances, rather than directly via the k8s API
3. Refactors the UserSignup create functions into a more flexible API

Fixes:

- https://issues.redhat.com/browse/CRT-1095
- https://issues.redhat.com/browse/CRT-945

### CODE REVIEW

File: test/e2e/user_management_test.go
Comment (alexeykazakov) [inline test/e2e/user_management_test.go:335]: Why not using ` s.newUserRequest()` for that?

File: test/e2e/user_management_test.go
Comment (alexeykazakov) [inline test/e2e/user_management_test.go:415]: Maybe also worth checking/waiting for the MUR to be deleted? Or it's already covered in other tests?

File: testsupport/user_request.go
Comment (alexeykazakov) [inline testsupport/user_request.go:200]: Shouldn't we check that the error is 409 (conflict)? And fail right away if it is not?
And we actually already have `UpdateUserSignupSpec()` function which you can use to update the UserSignup and which will deal with 409 conflicts: https://github.com/codeready-toolchain/toolchain-e2e/blob/master/testsupport/wait/host.go#L147

**Review by alexeykazakov — APPROVED** [review]
I like the new approach with Signup requests!
I have a few comments though. Please take a look.

**Review by alexeykazakov — APPROVED** [review]
I like the new approach with Signup requests!
I have a few comments though. Please take a look.

File: test/e2e/user_management_test.go
Comment (sbryzak) [reply test/e2e/user_management_test.go:336]: I wrote this test before I did the other refactoring, but yes I should be using the new API.  Fixed.

File: test/e2e/user_management_test.go
Comment (sbryzak) [reply test/e2e/user_management_test.go:416]: It's covered in other tests, but I think it's a good idea to confirm here also since it is supposed to be a full deactivation lifecycle test.  Fixed.

File: testsupport/user_request.go
Comment (sbryzak) [reply testsupport/user_request.go]: Ah that's a useful function... I've updated to use that instead.

**Review by MatousJobanek — COMMENTED** [review]
You should clean up all created UserSignups at the end of each test/sub-test. Something similar to what is done here: https://github.com/codeready-toolchain/toolchain-e2e/blob/master/testsupport/user_setup.go#L119-L123

In addition - you didn't modify all the current creation of the UserSignups.
Also - the new ` testsupport/signup_request.go` file has some overlap with the already existing [user_setup.go](https://github.com/codeready-toolchain/toolchain-e2e/blob/master/testsupport/user_setup.go) - could you please merge it?

BTW - as for the cleanup  - in the case of the failing test it's a bit tricky because the `cleanup()` functions are called at the end of the suite, not at the end of each of the tests. I faced the same problem in my current PR - you can see how I solved it: 
the cleanup: https://github.com/codeready-toolchain/toolchain-e2e/blob/7503d69cc61e030a1409a9dc793d9dfa8d4d4d99/test/e2e/usersignup_test.go#L32-L37
the new logic:  https://github.com/codeready-toolchain/toolchain-e2e/blob/7503d69cc61e030a1409a9dc793d9dfa8d4d4d99/testsupport/wait/awaitility.go#L468-L509

Please, take a look at the whole PR https://github.com/codeready-toolchain/toolchain-e2e/pull/325 I believe that you could reuse the clean logic when the PR is merged

**Comment by sbryzak** [comment]
Hey @MatousJobanek , thanks for the tips regarding cleanup.  I've added the cleanup steps to the new `signup_request.go`, plus tried adding the post-test cleanup as you suggested (adding the `Cleanup()` function to awaitility etc plus calling the `Clean()` function in `TearDownTest()` as you suggested, however while the UserSignup is now created successfully the test is still failing on the next line:

```
     usersignup_test.go:166: 
        	Error Trace:	usersignup_test.go:166
        	Error:      	Should not be: "member-ci-op-rfhd01ft-ceea8.origin-ci-int-aws.dev.rhcloud.com"
        	Test:       	TestRunUserSignupIntegrationTest/TestProvisionToOtherClusterWhenOneIsFull 
```

It seems that this test is making certain implicit assumptions as to the state of the cluster in order for it to pass successfully.  Instead of making these assumptions, shouldn't the test be confirming that the clusters are in the necessary state (in the `// given` section of the test) before the test is carried out?  At the very least it should throw a descriptive error detailing why the clusters aren't in the required state.

**Comment by sbryzak** [comment]
Also, for the tests that use `CreateAndApproveSignup()` from `user_setup.go`, I'm not going to refactor those in this PR.  The goal with this PR was to cease creating `UserSignup` instances directly via the k8s API, and the `CreateAndApproveSignup()` function doesn't use the API.  I agree there is some overlap, but I'd rather get this PR merged first in order to address the primary goal and not let its size blow out, then will deal refactoring to reduce overlap as a separate issue.

**Comment by MatousJobanek** [comment]
>I've added the cleanup steps to the new signup_request.go, plus tried adding the post-test cleanup as you suggested (adding the Cleanup() function to awaitility etc plus calling the Clean() function in TearDownTest() as you suggested, however while the UserSignup is now created successfully the test is still failing

I mentioned it in the review - you copied the code but never used it. The cleanup that you actually added doesn't wait until the UserSignup is completely deleted which causes test flakiness. 

> It seems that this test is making certain implicit assumptions as to the state of the cluster in order for it to pass successfully. Instead of making these assumptions, shouldn't the test be confirming that the clusters are in the necessary state (in the // given section of the test) before the test is carried out? 

The only assumption the test is making is that it expects that all previous tests are completely finished (no matter with which result/state). There should be no provisioning/deletion going on from the previous tests.
The test failure are caused because:
* you don't wait until the MUR is completely provisioned in the previous test
* you don't wait until the UserSignups are completely deleted

As you can see from the test setup https://github.com/codeready-toolchain/toolchain-e2e/blob/d98b3edf61bf68276df2ef7a9f8e6aa7f30507e0/test/e2e/usersignup_test.go#L136-L145 the test doesn't expect any specific number of the provisioned users, it just expects that the numbers are final.

> Also, for the tests that use CreateAndApproveSignup() from user_setup.go, I'm not going to refactor those in this PR. The goal with this PR was to cease creating UserSignup instances directly via the k8s API, and the CreateAndApproveSignup() function doesn't use the API. I agree there is some overlap, but I'd rather get this PR merged first in order to address the primary goal and not let its size blow out, then will deal refactoring to reduce overlap as a separate issue.

ok, fair enough. Just please don't forget to do that. I'm saying that because there is still one old review that hasn't been addressed yet https://github.com/codeready-toolchain/host-operator/pull/393#pullrequestreview-650157648

File: test/e2e/user_management_test.go
Comment (sbryzak) [reply test/e2e/user_management_test.go:513]: Hey Matous,  this PR was done in two stages; the first was to introduce the SignupRequest API to replace `createAndCheckUserSignup`, and all the refactorings for this were done first.  As a result of the refactoring, a number of inconsistencies in the previous tests now manifested as test failures, as a result of `UserSignup` resource instances being manually created with a state that was not in accordance with the host config.  In the process of fixing these issues I probably missed updating a few of the comment details.  The second stage of refactoring was to introduce `ConditionSet` and more fine-grained condition functions, the existing functions were getting out of control with the various combinations of `Complete` / `Approved` status conditions.  I've now updated the comment here to be more accurate.

File: test/e2e/user_workloads_test.go
Comment (sbryzak) [reply test/e2e/user_workloads_test.go:47]: Added call here to `EnsureMUR()`.

File: test/e2e/usersignup_test.go
Comment (sbryzak) [reply test/e2e/usersignup_test.go:49]: Added call here to `EnsureMUR()` also.

File: testsupport/signup_request.go
Comment (sbryzak) [reply testsupport/signup_request.go:233]: I've added a `WaitUntilUserSignupDeleted()` function to ensure the `UserSignup` is deleted, and set the propagation policy as suggested.  I'll also ensure the `Cleanup()` function is getting called.

File: testsupport/wait/awaitility.go
Comment (sbryzak) [reply testsupport/wait/awaitility.go:476]: Ah you're right, I had thought this was called as part of the call to `Clean()` but I now see that it's not.  Looking closer at your other PR, do I actually need this function?  I see that you are using it in [`CreateWithCleanup()`](https://github.com/codeready-toolchain/toolchain-e2e/pull/325/files#diff-4abc90985c7fe1e595e0d6f5131e1f4757a6f00c9239f40ae09375d12f15e530R466) but this is for creating resources directly via the client, which `SignupRequest` doesn't do.  Won't the following code be sufficient?  Based on the docs, this should be invoked after the test is complete:

> // Cleanup registers a function to be called when the test and all its
// subtests complete. Cleanup functions will be called in last added,
// first called order.


```
	// We also need to ensure that the UserSignup is deleted at the end of the test (if the test itself doesn't delete it)
	r.t.Cleanup(func() {
		propagationPolicy := metav1.DeletePropagationForeground
		opts := client.DeleteOption(&client.DeleteOptions{
			PropagationPolicy: &propagationPolicy,
		})

		// Delete the UserSignup
		if err := r.hostAwait.Client.Delete(context.TODO(), r.userSignup, opts); err != nil && !errors.IsNotFound(err) {
			require.NoError(r.t, err)
		}

		// Ensure the UserSignup has been deleted
		require.NoError(r.t, r.hostAwait.WaitUntilUserSignupDeleted(r.userSignup.Name))
	})
```

File: testsupport/signup_request.go
Comment (sbryzak) [reply testsupport/signup_request.go:73]: This is actually the behaviour of the [State manager](https://github.com/codeready-toolchain/toolchain-common/blob/master/pkg/states/state_manager.go#L13).  Manually setting `approved` to true removes the `verification-required` state, this is something that Alexey and I both discussed and agreed on previously.

File: testsupport/signup_request.go
Comment (sbryzak) [reply testsupport/signup_request.go:156]: Sure thing, that's a good idea.  Fixed.

File: testsupport/signup_request.go
Comment (sbryzak) [reply testsupport/signup_request.go:116]: What else would this new type return besides the resources?

File: testsupport/signup_request.go
Comment (sbryzak) [reply testsupport/signup_request.go:186]: The test might still pass if during user provisioning the correct cluster happens to be randomly selected, however if not, and the test is actually checking for which cluster the user was provisioned in (like some tests do, which are the reason I added this error message) then it would fail, and with a much more cryptic error.

File: testsupport/signup_request.go
Comment (sbryzak) [reply testsupport/signup_request.go:208]: If none of the conditions (`verificationRequired`, `manuallyApprove` or `targetCluster`) are set/true, then what would the `UpdateUserSignupSpec` call be updating?

**Comment by sbryzak** [comment]
@MatousJobanek thanks for the great review, this has made clearer some of the unknowns around the `TestProvisionToOtherClusterWhenOneIsFull()` test, and your other suggestions were appreciated also.

> The only assumption the test is making is that it expects that all previous tests are completely finished (no matter with which result/state). There should be no provisioning/deletion going on from the previous tests.
> The test failure are caused because:
> 
>     you don't wait until the MUR is completely provisioned in the previous test
>     you don't wait until the UserSignups are completely deleted
> 

Is there something we can query to validate this?  If the test requires that the cluster/s is "stable" and not still reconciling leftovers from previous tests then I wonder if there is some k8s API to detect this.  Failing that, we should add some detailed comments at the start of this that describe these prerequisites.

File: testsupport/conditions.go
Comment (sbryzak) [reply testsupport/conditions.go:12]: I'm very shortly going to be creating a new Epic with a number of suggested improvements to the e2e tests, let's create an issue for this and we can examine how we can improve it further.

File: test/e2e/user_management_test.go
Comment (MatousJobanek) [reply test/e2e/user_management_test.go]: Ok, I understand that.
thanks for updating it :+1:

File: testsupport/wait/awaitility.go
Comment (MatousJobanek) [reply testsupport/wait/awaitility.go:482]: Unfortunately not.
In my PR I figured out that there are some MURs still present & being deleted also when the UserSignup is gone. I have no idea why.
I'm struggling with the same issues as well, but I hope that I found the solution/workaround. 
As I see it - the easiest for you will be to wait until I merge the PR with the improved cleanup and then to reuse it.

File: testsupport/conditions.go
Comment (MatousJobanek) [reply testsupport/conditions.go:12]: cool, thanks :+1:

File: testsupport/signup_request.go
Comment (MatousJobanek) [reply testsupport/signup_request.go:76]: I know, I'm just wondering if these additional checks are really needed in e2e tests. Does it bring any benefit?
I would expect that I can configure any crazy combination of the values in e2e tests and test the controller/operator with it.

File: testsupport/signup_request.go
Comment (MatousJobanek) [reply testsupport/signup_request.go:122]: probably nothing else for now
But it will bring better understandability to the code.

File: testsupport/signup_request.go
Comment (MatousJobanek) [reply testsupport/signup_request.go:178]: ok, fair enough :+1:

File: testsupport/signup_request.go
Comment (MatousJobanek) [reply testsupport/signup_request.go:200]: nothing, but do we care? It doesn't really matter if anything is updated, right? But the code will be much cleaner.

**Comment by MatousJobanek** [comment]
> Is there something we can query to validate this? If the test requires that the cluster/s is "stable" and not still reconciling leftovers from previous tests then I wonder if there is some k8s API to detect this. 

you could use metrics for that, but I wouldn't rely on it - there can be always something to be reconciled. 

> Failing that, we should add some detailed comments at the start of this that describe these prerequisites.

yup, feel free to add it :wink: - if not, then I can try my best :-)

File: testsupport/wait/awaitility.go
Comment (MatousJobanek) [reply testsupport/wait/awaitility.go:482]: done - the PR is merged. You can delete the copy-pasted code and reuse the `Cleanup` functions from the commit

File: testsupport/signup_request.go
Comment (sbryzak) [reply testsupport/signup_request.go:200]: I'd prefer not to make an update call if it's not required, especially if it's just for the sake of making the code appear cleaner.

File: testsupport/wait/awaitility.go
Comment (sbryzak) [reply testsupport/wait/awaitility.go]: Looks like your PR has fixed some things, as the tests are now passing here also.

File: testsupport/signup_request.go
Comment (sbryzak) [reply testsupport/signup_request.go:122]: I'd prefer not to introduce a new type here, it doesn't seem to bring any benefit.

File: testsupport/signup_request.go
Comment (sbryzak) [reply testsupport/signup_request.go:76]: The state manager is intended to enforce a set of rules for the sake of the consistent management of UserSignup states.  There's no additional checks here in the actual e2e test, the comment on the `VerificationRequired()` function is simply there to draw attention to the fact of the state manager's behaviour in this instance.  I don't want to start bypassing the state manager and manipulating the `state` values directly, these rules are enforced by our controllers so I don't see the point in breaking those rules in our tests.

File: testsupport/signup_request.go
Comment (MatousJobanek) [inline testsupport/signup_request.go:234]: @sbryzak You forgot to remove this code and replace it with the `Awaitility.Cleanup` function. Could you please fix that?

File: testsupport/signup_request.go
Comment (MatousJobanek) [reply testsupport/signup_request.go:76]: ah, yeah - I forgot that you use state manager in e2e tests as well. Sorry for confusions :wink:

**Pull request merged**
